File: python/library/aliyun/signature.py
from config.aliyun import Aliyun
from util.util import Util
from util.base64 import Base64
from util.hash import Hash

# 签名
class Signature:

  # 签名直传
  def PolicySign(expireTime: int, maxSize: int):
    # 配置
    cfg = Aliyun.RAM()
    conditions = []
    # 限制大小
    conditions += [['content-length-range', 0, maxSize]]
    # 超时时间
    now = Util.Time()
    expire = now + expireTime
    expiration = Util.GmtISO8601(expire)
    # 签名数据
    policyStr = Util.JsonEncode({'expiration':expiration, 'conditions':conditions})
    policy = Base64.ToStr(Base64.Encode(Base64.ToByte(policyStr)))
    signature = Base64.ToStr(Base64.Encode(Hash.HmacSha1(policy, Base64.ToByte(cfg['AccessKeySecret']))))
    # 返回
    return {
      'accessid': cfg['AccessKeyId'],
      'policy': policy,
      'signature': signature,
      'expire': expire,
    }

File: java/src/main/java/webmis/library/aliyun/Signature.java
package webmis.library.aliyun;

import java.util.HashMap;

import com.alibaba.fastjson.JSONArray;
import com.alibaba.fastjson.JSONObject;

import webmis.config.Aliyun;
import webmis.service.Base;
import webmis.util.Base64;
import webmis.util.Hash;
import webmis.util.Util;

/* 签名 */
public class Signature extends Base {

  /* 签名直传 */
  public static JSONObject PolicySign(long expireTime, long maxSize) {
    // 配置
    HashMap<String, Object> cfg = Aliyun.RAM();
    JSONArray conditions = new JSONArray();
    JSONArray tmp = new JSONArray();
    // 限制大小
    tmp.add("content-length-range");
    tmp.add(0);
    tmp.add(maxSize);
    conditions.add(tmp);
    // 超时时间
    long now = Util.Time();
    long expire = now + expireTime;
    String expiration = Util.GmtISO8601(expire);
    // 签名数据
    JSONObject policyData = new JSONObject();
    policyData.put("expiration", expiration);
    policyData.put("conditions", conditions);
    String policyStr = Util.JsonEncode(policyData);
    String policy = Base64.ToStr(Base64.Encode(Base64.ToByte(policyStr)));
    String signature = Hash.Base64Encode(Hash.HmacSha1(policy, Base64.ToByte(cfg.get("AccessKeySecret").toString())));
    // 返回
    JSONObject res = new JSONObject();
    res.put("accessid", cfg.get("AccessKeyId"));
    res.put("policy", policy);
    res.put("policy", policy);
    res.put("signature", signature);
    res.put("expire", expire);
    return res;
  }
  
}
